controllers/restore.go

```diff
@@ -839,3 +845,164 @@ func setUserRestoreFilters(
+
+	reqLogger := log.FromContext(ctx)
+	reqLogger.Info("get cluster-backup  internalhubcomponent")
+
+	mchGVRList := schema.GroupVersionResource{Group: ihcGroup,
+		Version: "v1", Resource: "internalhubcomponents"}
+
+	mchList, err := dyn.Resource(mchGVRList).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return &unstructured.Unstructured{}, err
+	}
+	//
+	reqLogger.Info("get cluster-backup  internalhubcomponent")
+	for i := range mchList.Items {
+		item := mchList.Items[i]
+
+		if item.GetName() == "cluster-backup" {
+			return &item, nil
+		}
+	}
+
+	return &unstructured.Unstructured{}, err
+
+}
+
+// remove the acm finalizer and
+// remove internalhubcomponent finalizer if this is the only restore resource found
+func removeResourcesFinalizer(
+	ctx context.Context,
+	c client.Client,
+	dyn dynamic.Interface,
+	internalHubResource *unstructured.Unstructured,
+	acmRestore *v1beta1.Restore,
+) error {
+
+	reqLogger := log.FromContext(ctx)
+
+	// Remove restore finalizer. Once all finalizers have been
+	// removed, the object will be deleted.
+	needsUpdate := controllerutil.RemoveFinalizer(acmRestore, acmRestoreFinalizer)
+
+	// if no other restore resources, remove mch finalizer
+	acmRestoreList := v1beta1.RestoreList{}
+	if err := c.List(
+		ctx,
+		&acmRestoreList,
+		client.InNamespace(acmRestore.GetNamespace())); err == nil &&
+		len(acmRestoreList.Items) == 1 {
+
+		// remove InternalHubResource restore finalizer if this is the last resource to be deleted
+		mchGVRList := schema.GroupVersionResource{
+			Group:   ihcGroup,
+			Version: "v1", Resource: "internalhubcomponents"}
+		if internalHubResource != nil {
+			if fins := internalHubResource.GetFinalizers(); fins != nil && findValue(fins, acmRestoreFinalizer) {
+				fins = remove(fins, acmRestoreFinalizer)
+				internalHubResource.SetFinalizers(fins)
+				//save internal hub resource
+				if _, err := dyn.Resource(mchGVRList).Namespace(internalHubResource.GetNamespace()).Update(ctx,
+					internalHubResource, metav1.UpdateOptions{}); err != nil && !errors.IsNotFound(err) {
+					// ignore not found error
+					reqLogger.Info(err.Error())
+					return err
+				}
+			}
+		}
+	}
+
+	if needsUpdate {
+		if err := c.Update(ctx, acmRestore); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// add the acm finalizer for the restore resource
+// and if needed for the internalhubcomponent
+func addResourcesFinalizer(
+	ctx context.Context,
+	c client.Client,
+	dyn dynamic.Interface,
+	internalHubResource *unstructured.Unstructured,
+	acmRestore *v1beta1.Restore,
+) error {
+
+	reqLogger := log.FromContext(ctx)
+
+	// add the finalizer for the internalhubcomponent
+	if internalHubResource != nil &&
+		internalHubResource.GetDeletionTimestamp() == nil {
+		// process internalhubcomponent, it is not marked for deletion
+		reqLogger.Info("adding finalizer to internalhubcomponent")
+		needsUpdate := false
+		resFins := internalHubResource.GetFinalizers()
```

Comment:
Possible to use the controllerutil function here to addFinalizer?

Reply:
internalhubcomponent is an unstructure resource, there is no addFinalizer; just get/setFinalizers
I am getting the finalizer list, and I add the restores.cluster.open-cluster-management.io/finalizer if not set